benchmarks/pmap_benchmark.py

```diff
@@ -0,0 +1,144 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""To run on CPU with 500 CPU devices:
+
+CUDA_VISIBLE_DEVICES= XLA_FLAGS=--xla_force_host_platform_device_count=500 python3 pmap_benchmark.py
+
+To make it run faster, set env var TARGET_TOTAL_SECS to a low number (e.g. 2).
+"""
+import os
+import time
+
+import numpy as onp
+from tabulate import tabulate
+
+import jax
+import jax.numpy as np
+from jax import pmap
+from jax.util import safe_zip
+
+def pstd(x):
+  return x.std() / x.mean() * 100
+
+
+def benchmark(f, iters=None, warmup=None, name=None, target_total_secs=None):
+  if target_total_secs is None:
+    target_total_secs = int(os.getenv("TARGET_TOTAL_SECS", 10))
+
+  if warmup is None:
+    if iters is None:
+      warmup = 1
+    else:
+      warmup = onp.clip(1, iters // 10, 10)
+  for _ in range(warmup):
+    f()
+
+  times = []
+  count = 0
+  while (count < iters if iters is not None
+         else sum(times) < target_total_secs):
+    start = time.time()
+    f()
+    end = time.time()
+    times.append(end - start)
+    count += 1
+
+  times = onp.array(times)
+  print("---------Benchmark results for %s---------" % (name or f.__name__))
+  print("mean=%f std=%f %%std=%f total=%f" %
+        (times.mean(), times.std(), pstd(times), times.sum()))
+  print("#iters=%d #warmup=%d" % (count, warmup))
+  print()
+  return times
+
+
+def benchmark_suite(funcs, params_list, param_names, name,
```

Comment:
I think it would be useful to document this function, because I do not think it is obvious how to use it. It would also be useful to move it into a benchmarking library.

If I understand correctly, the `funcs` are already specialized to the `param_list`, in that exact order, but the latter are passed in along with `param_names` only for printing purposes. Wouldn't the following alternate API be easier to use, because it would encapsulate the specialization of the benchmark function to the list of parameters, along with the 
construction of the specialized function name?

```
benchmark_suite(prepare: Callable[[Dict], Callable], params_list: List[Dict], name: str):
 """Benchmarks a function for several combinations of parameters.

 Args:
    prepare: given kwargs returns a benchmark function specialized to the kwargs. 
    params_list: a list of kwargs on which to run the benchmark.
    name: the name of this benchmark suite
 """ 
```

Reply:
I moved `benchmark` and `benchmark_suite` into a new `benchmark.py` file.

Good idea with the simplified API, done. Note that I take dicts but sort them alphabetically so they print in a consistent order. We could do something fancier in the future if we want (e.g. introspect the function argument order), but I think this is ok for now. I thought doing the sorting automatically is better than requiring the caller to use consistent OrderedDicts.